Change spanmetrics filtering with multiple policies (#3308)
grafana/tempo: docs/sources/tempo/metrics-generator/span_metrics.md

### Filtering

In some cases, you may want to reduce the number of metrics that are produced

> from the `spanmetrics` processor by using a filter. With a filter you can

Reviewer: ```suggestion
from the `spanmetrics` processor by using a filter. With a filter, you can
```

> In some cases, you may want to reduce the number of metrics that are produced
> from the `spanmetrics` processor by using a filter. With a filter, you can

Reviewer: Moving "use a filter" to the front of the sentence because it's the method people can use to reduce the number of metrics.
```suggestion
In some cases, you may want to use a filter to reduce the number of metrics that are produced
from the `spanmetrics` processor. With a filter, you can
```

- `SPAN_KIND_PRODUCER`
- `SPAN_KIND_CONSUMER`

#### Filter policies

Each filter policy is evaluated for every span until an `include` policy is

> matched without a corresponding `exclude` policy match within the same filter.

Reviewer: Much better! Thank you.

> Each filter policy is evaluated in order for every span to determine if the
> span should be included for metrics.  Both `include` and `exclude` policies are
> used to *reject* spans from the metrics.  With no filter polices applied, all
> spans are included.

Reviewer: I'm having some trouble parsing the first sentence. Do you mean 1) that every filter policy is evaluated in a particular order (i.e., policy A first, policy B, policy C, etc.) or 2) that each policy is used to evaluated every span to determine whether a span should be included for metrics? 
```suggestion
Each filter policy is evaluated in order for every span to determine if the
span should be included for metrics.  Both `include` and `exclude` policies are
used to *reject* spans from the metrics.  With no filter polices applied, all
spans are included.
```

Reviewer: Okay, reading later down I think you're referring to the order of the filter policies.

Author: Okay I've updated.  The policies are evaluated in order yes (0,1,2), but only until there is a match to include without a matching exclude on the same policy.  Hows that read?

{{% admonition type="note" %}}

> An exclude match will take precedence over an include match within the *same*
> policy, but not later polices.
> {{% /admonition %}}
> 
> {{% admonition type="note" %}}
> A matching exclude of one policy will be ignored if a later policy `include`
> matches, and that same policy does not also have a matching `exclude` for the
> span.
> {{% /admonition %}}

Reviewer: Changing future tense to present tense to follow our [style guidelines](https://grafana.com/docs/writers-toolkit/write/style-guide/style-conventions/#write-in-present-tense). 

```suggestion
An exclude match takes precedence over an include match within the *same*
policy, but not later polices.
{{% /admonition %}}

{{% admonition type="note" %}}
A matching exclude of one policy is ignored if a later policy `include`
matches, and that same policy does not also have a matching `exclude` for the
span.
{{% /admonition %}}
```

In the above, we first include all spans which have a `resource.location` that begins with `eu-` with the `include` statement, and then exclude those with begin with `dev-`.
In this way, a flexible approach to filtering can be achieved to ensure that only metrics which are important are generated.

Additionally, when multiple filter policies are are used, any policy that is

> matched will result in the span being included in the metric export.

Reviewer: ```suggestion
matched results in the span being included in the metric export.
```

```yaml
---
metrics_generator:
  processor:
    span_metrics:
      filter_policies:
        - include:
            match_type: regex
            attributes:
              - key: resource.location
                value: eu-.*
          exclude:
            match_type: regex
            attributes:
              - key: resource.tier
                value: dev-.*
        - include:
            match_type: regex
            attributes:
              - key: resource.location
                value: sa-.*
              - key: resource.platform
                value: k8s-.*
          exclude:
            match_type: regex
            attributes:
              - key: resource.tier
                value: dev-.*
```

> In this example, spans that match a location beginning with `eu-` or `sa-` will
> be included so long as they are not a `resource.tier` beginning with `dev-`.
> This example could also be written with a single regex expression, except that
> in the second policy we have an additional constraint to match against.  If
> either the `resource.location` or the `resource.platform` is not matched in the
> second policy, then the span is not included.

Reviewer: ```suggestion
In this example, spans that match a location beginning with `eu-` or `sa-` are
included so long as they are not a `resource.tier` beginning with `dev-`.
This example could also be written with a single regex expression, except that
in the second policy, we have an additional constraint to match against. If
either the `resource.location` or the `resource.platform` is not matched in the
second policy, then the span is not included.
```

Author: I reworked this a bit.

            attributes:
              - key: resource.location
                value: eu-.*
          exclude:
            match_type: regex
            attributes:
              - key: resource.tier
                value: dev-.*
```

> In the above, we first include all spans which have a `resource.location` that
> begins with `eu-` with the `include` statement, and then `exclude` those with
> begin with `dev-`. In this way, a flexible approach to filtering can be
> achieved to store only those metrics which are important.
> 

Reviewer: ```suggestion
The above example first includes all spans which have a `resource.location` that
begins with `eu-` with the `include` statement, and then `exclude` those with
begin with `dev-`. In this way, you can create a flexible approach to filtering to store 
only those metrics which are important.

```

Additionally, when multiple filter policies are are used, any policy that is
matched results in the span being included in the metric export.

```yaml
---
metrics_generator:
  processor:
    span_metrics:
      filter_policies:
        - include:
            match_type: regex
            attributes:
              - key: resource.location
                value: eu-.*
          exclude:
            match_type: strict
            attributes:
              - key: resource.tier
                value: research
        - include:
            match_type: regex
            attributes:
              - key: resource.location
                value: sa-.*
              - key: resource.platform
                value: k8s-.*
          exclude:
            match_type: regex
            attributes:
              - key: resource.tier
                value: research
```

> In this example, spans that match a location beginning with `eu-` or `sa-` will
> be included, but in both cases we also exclude spans which match

Reviewer: ```suggestion
In this example, spans that match a location beginning with `eu-` or `sa-` are 
included, but both cases also exclude spans which match
```

`{resource.tier="research"}`.

> This example could also be written with a single regex expression, except that
> in the second policy we have an additional constraint to match against for
> `k8s-.*`.  If either the `resource.location` or the `resource.platform` is not
> matched in the second policy, then the span is not included.

Reviewer: Slight tweak to exclude second person "we". 
```suggestion
This example could also be written with a single regex expression, except that
in the second policy has an additional constraint to match against for
`k8s-.*`.  If either the `resource.location` or the `resource.platform` is not
matched in the second policy, then the span is not included.
```
